### PR TITLE
nspawn: shorten ifname if the prefixed ifname exceeds the length limit

### LINKED ISSUE #28844: systemd-nspawn: MACVLAN network interface, interface name not valid
### systemd version the issue has been seen with

254.1-1

### Used distribution

Arch

### Linux kernel version used

6.4.10

### CPU architectures issue was seen on

x86_64

### Component

systemd-nspawn

### Expected behaviour you didn't see

_No response_

### Unexpected behaviour you saw

I'm getting an error message from systemd-nspawn:
MACVLAN network interface, interface name not valid: mv-enp0s20f0u2c2

The interface does exist (USB ethernet adapter), "ip address":
3: enp0s20f0u2c2: <BROADCAST,MULTICAST,UP,LOWER_UP> mtu 1500 qdisc fq_codel state UP group default qlen 1000

systemd-nspawn parameter used:
--network-macvlan=enp0s20f0u2c2

There is another network interface "eno2", which does work. "enp0s20f0u2c2" was working with systemd 253.

### Steps to reproduce the problem

_No response_

### Additional program output to the terminal or log subsystem illustrating the issue

_No response_

### PR DESCRIPTION
**Pull request by yuwata**
Follow-up for 2f091b1b49543aade4aad9ec3b35b3665abac3e7.

Fixes #28844.

### CODE REVIEW

File: src/nspawn/nspawn-network.c [hunk 1]
```diff
@@ -775,15 +776,19 @@ static int network_iface_pair_parse(const char* iftype, char ***l, const char *p
...
-                        else
+                        } else
                                 b = strdup(a);
                 } else
                         b = strdup(interface);
                 if (!b)
                         return log_oom();
 
+                if (may_shorten)
+                        shorten_ifname(b);
```
Comment (mrc0mmand) [inline src/nspawn/nspawn-network.c]:
```suggestion
                        (void) shorten_ifname(b);
```

**Review by mrc0mmand — APPROVED** [review]
Could you add a test case for this? I think simple

```diff
diff --git a/test/units/testsuite-13.nspawn.sh b/test/units/testsuite-13.nspawn.sh
index 9bcdec9298..8d7a2e92ab 100755
--- a/test/units/testsuite-13.nspawn.sh
+++ b/test/units/testsuite-13.nspawn.sh
@@ -341,7 +341,7 @@ testcase_nspawn_settings() {
     rm -f "/etc/systemd/nspawn/$container.nspawn"
     mkdir -p "$root/tmp" "$root"/opt/{tmp,inaccessible,also-inaccessible}
 
-    for dev in sd-host-only sd-shared{1,2} sd-macvlan{1,2} sd-ipvlan{1,2}; do
+    for dev in sd-host-only sd-shared{1,2} sd-macvlan{1,2} sd-macvlanloong sd-ipvlan{1,2} sd-ipvlanlooong; do
         ip link add "$dev" type dummy
     done
     udevadm settle
@@ -395,8 +395,8 @@ VirtualEthernet=yes
 VirtualEthernetExtra=my-fancy-veth1
 VirtualEthernetExtra=fancy-veth2:my-fancy-veth2
 Interface=sd-shared1 sd-shared2:sd-shared2
-MACVLAN=sd-macvlan1 sd-macvlan2:my-macvlan2
-IPVLAN=sd-ipvlan1 sd-ipvlan2:my-ipvlan2
+MACVLAN=sd-macvlan1 sd-macvlan2:my-macvlan2 sd-macvlanloong
+IPVLAN=sd-ipvlan1 sd-ipvlan2:my-ipvlan2 sd-ipvlanlooong
 Zone=sd-zone0
 Port=80
 Port=81:8181
```

should do the trick.

**Comment by mrc0mmand** [comment]
Also, the original paths did add an alternative name to the shortened interface with the original name, which now doesn't happen. Is that a problem?

**Comment by yuwata** [comment]
Updated, and applied the suggested test change. PTAL.

File: src/nspawn/nspawn-network.c [hunk 2]
```diff
@@ -774,20 +774,22 @@ static int network_iface_pair_parse(const char* iftype, char ***l, const char *p
...
 
-                if (isempty(interface)) {
-                        if (ifprefix)
-                                b = strjoin(ifprefix, a);
-                        else
-                                b = strdup(a);
-                } else
+                /* Here, we only check the validity of the specified second name. If it is not specified,
+                 * the copied or prefixed name should be already valid, except for its length. If it is too
+                 * long, then it will be shorten later. */
```
Comment (poettering) [inline src/nspawn/nspawn-network.c]: shorten → shortened

**Review by mrc0mmand — APPROVED** [review]

**Comment by keszybz** [comment]
LGMT>

**Pull request merged**